src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,829 @@
+  query: string
+  matchingLineNumbers: number[]
+}
+
+// Debounces the search handler which ensures that we don’t trigger several searches while the user is still typing.
+const debouncedHandleSearchInputValue = debounce(handleSearchInputValue, 150)
+
+const props = defineProps({
+  /**
+   * ID value used for form elements like the search input and its label.
+   */
+  id: {
+    type: String,
+    required: false,
+    default: 'code-block',
+  },
+
+  /**
+   * The code that will be rendered as a text node inside of a `<pre><code></code></pre>` fragment.
+   */
+  code: {
+    type: String,
+    required: true,
+  },
+
+  /**
+   * The syntax language of `props.code` (e.g. `'json'`).
+   */
+  language: {
+    type: String,
+    required: true,
+  },
+
+  /**
+   * Shows an actions bar with a search input and related action buttons.
+   */
+  isSearchable: {
+    type: Boolean,
+    required: false,
+    default: false,
+  },
+
+  /**
+   * Allows controlling the processing state from outside the component. This allows a parent component to show the processing icon when it’s, for example, currently syntax highlighting the code.
+   */
+  isProcessing: {
+    type: Boolean,
+    required: false,
+    default: false,
+  },
+
+  /**
+   * Used as the initial value of the search input. Can be used to initialize a code block with a query which was read from client storage.
+   */
+  query: {
+    type: String,
+    required: false,
+    default: '',
+  },
+
+  /**
+   * Controls whether to show a copy button. It copies the original code (i.e. the value of `props.code`) to the clipboard.
+   */
+  showCopyButton: {
+    type: Boolean,
+    required: false,
+    default: true,
+  },
+
+  /**
+   * Controls whether to show line numbers.
+   */
+  showLineNumbers: {
+    type: Boolean,
+    required: false,
+    default: true,
+  },
+})
+
+const emit = defineEmits<{
+  (event: 'highlight', data: CodeBlockEventData): void
+  (event: 'matching-lines-change', data: CodeBlockEventData): void
+  (event: 'query-change', data: string): void
+}>()
+
+const query = ref(props.query)
+const isProcessingInternally = ref(false)
+const isRegExpMode = ref(false)
+const isFilterMode = ref(false)
+const regExpError = ref<Error | null>(null)
+const codeBlock = ref<HTMLElement | null>(null)
+const numberOfMatches = ref(0)
+const matchingLineNumbers = ref<number[]>([])
+const currentLineIndex = ref<null | number>(null)
+
+const totalLines = computed(() => Array.from({ length: props.code.split('\n').length }, (_, index) => index + 1))
+const linePrefix = computed(() => props.id.toLowerCase().replace(/\s+/g, '-'))
+const isProcessing = computed(() => props.isProcessing || isProcessingInternally.value)
+const isShowingFilteredCode = computed(() => isFilterMode.value && filteredCode.value !== '')
+const regExpButtonAppearance = computed(() => regExpError.value !== null ? 'danger' : isRegExpMode.value ? 'secondary' : 'outline')
+const filteredCode = computed(function() {
```

Comment:
Can you add return types to these?

Reply:
I can, but I shouldn’t. Computed properties aren’t functions (which should always be type-annotated) and in the presence of correctly typed sources, a computed property will always automatically infer its correct type already. All computed properties as they’re currently used in the code block’s implementation correctly infer there type with the type of `filteredCode` being `string`.